File: sensor.py
import time
import mysql.connector
import Adafruit_DHT
from datetime import datetime

# MySQL database connection
conn = mysql.connector.connect(
    host="localhost",
    user="your_username",  # Replace with your MySQL username
    password="your_password",  # Replace with your MySQL password
    database="sensor_data"  # Replace with your database name
)
cursor = conn.cursor()

# Create table if not exists
cursor.execute('''
    CREATE TABLE IF NOT EXISTS temperature (
        id INT AUTO_INCREMENT PRIMARY KEY,
        timestamp DATETIME,
        temperature FLOAT
    )
''')
conn.commit()

# DHT sensor configuration
DHT_SENSOR = Adafruit_DHT.DHT11  # Change to DHT22 if needed
DHT_PIN = 2  # GPIO pin connected to DHT sensor

def store_temperature(temperature):
    # Get the current timestamp
    timestamp = datetime.now()

    # Insert the data into the MySQL table
    cursor.execute("INSERT INTO temperature (timestamp, temperature) VALUES (%s, %s)",
                   (timestamp, temperature))
    conn.commit()
    print(f"Stored data: {timestamp} - {temperature} °C")

def read_and_store_data():
    while True:
        # Read temperature and humidity from DHT sensor
        humidity, temperature = Adafruit_DHT.read(DHT_SENSOR, DHT_PIN)

        # Check if the temperature reading is valid
        if temperature is not None:
            print(f"Temperature: {temperature} °C")
            store_temperature(temperature)  # Store data into MySQL
        else:
            print("Failed to read temperature from DHT sensor!")

        # Wait before reading again
        time.sleep(5)  # Adjust the sleep time as needed

if __name__ == "__main__":
    read_and_store_data()

File: Sensor.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.time.LocalDateTime;
import java.util.Scanner;
import com.fazecast.jSerialComm.SerialPort;

public class Sensor {
    // MySQL Database credentials
    private static final String DB_URL = "jdbc:mysql://localhost:<port_no>/<DB_name>";
    private static final String DB_USER = "your_username"; 
    private static final String DB_PASSWORD = "your_password"; 

    public static void main(String[] args) {
        // Open the Serial Port
        SerialPort comPort = SerialPort.getCommPort("COM3"); 
        comPort.setBaudRate(9600);
        
        if (!comPort.openPort()) {
            System.out.println("Failed to open serial port!");
            return;
        }

        System.out.println("Listening for sensor data...");

        Scanner scanner = new Scanner(comPort.getInputStream());
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD)) {
            while (true) {
                // Read data from serial
                if (scanner.hasNextLine()) {
                    String data = scanner.nextLine().trim();
                    try {
                        float temperature = Float.parseFloat(data); // Convert received data to float
                        System.out.println("Received Temperature: " + temperature + "°C");
                        storeTemperature(conn, temperature); // Store data in MySQL
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid data received: " + data);
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            scanner.close();
            comPort.closePort();
        }
    }

    private static void storeTemperature(Connection conn, float temperature) {
        String sql = "INSERT INTO temperature (timestamp, temperature) VALUES (?, ?)";
        try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setObject(1, LocalDateTime.now()); // Current timestamp
            pstmt.setFloat(2, temperature);
            pstmt.executeUpdate();
            System.out.println("Stored in DB: " + temperature + "°C");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
